deepmd/dpmodel/utils/learning_rate.py

```
@@ -29,86 +29,274 @@ def __new__(cls: type, *args: Any, **kwargs: Any) -> Any:
+        )
+        # === Step 5. Compute decay_rate for exp scheduler ===
+        # Use decay_num_steps (num_steps - warmup_steps) for decay calculation
+        decay_total = self.decay_num_steps
         self.decay_steps = decay_steps
-        if self.decay_steps >= stop_steps:
-            self.decay_steps = default_ds
+
+        if self.decay_steps > decay_total:
+            raise ValueError(
+                f"decay_steps ({self.decay_steps}) must not exceed decay phase steps ({decay_total})."
+            )
```

Ruff (0.14.11)

Avoid specifying long messages outside the exception class

(TRY003)